### ISSUE TITLE
Abstract return type of callbacks

### ISSUE THREAD

**Issue by woeps** [sikanhe/reason-node#27]
I focused on the Stream bindings, but this probably holds true for other parts of this repo as well.

Currently the return type of a callback is `unit`. Therefore it's possible to just return any expression returning `unit`. To prevent missing callbacks, they could be typed to return an abstract type. The function passing in the callback would then expect this abstract type.
This way the compiler would know if you forgot to call the callback.

I used this approach in some adhoc bindings of streams and a parser and liked it. What are you thinking about this?

**Comment by austindd** [comment]
@woeps I apologize, but I am not quite sure I understand the problem you are describing. Could you provide some example code to demonstrate the issue?

**Comment by woeps** [comment]
Don't apologize! I should have provided an example in the first place.
Let's take [`ReasonNode.Stream.Transform.makeOptions`](https://github.com/sikanhe/reason-node/blob/master/src/Stream.re#L731) as an arbitary example.

If you simply change the return type of the `transform` and `flush` functions, it's not possible to *not* call the callback anymore. Like this:
```reason
type calledBackTransform;
type calledBackFlush;
[@bs.obj]
  external makeOptions:
    (
      ~objectMode: [@bs.as {json|false|json}] _,
      ~highWaterMark: int=?,
      ~emitClose: bool=?,
      ~autoDestroy: bool=?,
      ~transform: [@bs.this] (
                    (
                      t('w, 'r),
                      'w,
                      string,
                      (~err: option(Js.Exn.t), ~data: option('r)) => calledBackTransform
                    ) =>
                    calledBackTransform
                  ),
      ~flush: [@bs.this] (
                (
                  t('w, 'r),
                  (~err: option(Js.Exn.t), ~data: option('r)) => calledBackFlush
                ) =>
                calledBackFlush
              ),
      unit
    ) =>
    makeOptions('w, 'r);
```

This is not perfect, because I could "evilish" call `makeOptions` two or more times and use the callback provided by the other function. In this case the compiler won't complain and it's probably still  not the intended usage. But the chances of forgetting the callback ar less than before.

**Comment by sikanhe** [comment]
@woeps What do you mean by "missing callbacks" ? These callbacks are ones you pass to Nodejs, and Node internally calls it - Therefore I don't know what a "missing" callback mean in this scenario.

**Comment by woeps** [comment]
I'll try to explain on the example of the `~transform~` function, but the same holds true for `~flush` and similar functions in the node api.
The [node docs of `transform._transform`](https://nodejs.org/api/stream.html#stream_transform_transform_chunk_encoding_callback) state:
> The `callback` function **must** be called only when the current chunk is completely consumed.

Currently the type system will evaluate the `~transform` parameter successfully if it looks like this (minimal exampe): `~transform=[@bs.this](this, chunk, encoding, callback) => {let transformed_chunk = someOperation(chunk); Js.log(transformed_chunk}`
You can see the callback's call is missing, but since the function returns unit the types check.

Now if I introduce an abstract type like in my previous [example](https://github.com/sikanhe/reason-node/issues/27#issuecomment-625056715) then this wouldn't typecheck, since unit is not equal to the abstract type. The only way to satisfy the type constraint is to call the `callback` last: `~transform=[@bs.this](this, chunk, encoding, callback) => {let transformed_chunk = someOperation(chunk); Js.log(transformed_chunk}, callback(~error=None, ~data=transformed_chunk`.

**Comment by sikanhe** [comment]
@woeps Understand now, the callback we must call is the one given to us by the transform callback, and we want to encode that invariant into the type signature by using abstract type as the return type of both transform and the callback.

**Comment by sikanhe** [comment]
@woeps Do you think you can send a PR for this feature and see what the diff looks like?

**Comment by woeps** [comment]
@sikanhe sure, I will do so for the Stream module since that's what I'm most familiar with by now.

**Comment by austindd** [comment]
closed with #29
